File: dotnet/wwwroot/js/Kamanri/MyWebSocket.js

const IS_EFFECTIVE_CODE = 200;
const HEADER_LENGTH = 10;

const UINT32_LENGTH = 4;
const UINT8_LENGTH = 1;

const IS_EFFECTIVE_LOCATION = 0;
const MESSAGE_EVENT_CODE_LOCATION = 1;

const MESSAGE_TYPE_LOCATION = 5;
const LENGTH_LOCATION = 6;
const MESSAGE_LOCATION = 10;

var WebSocketMessageEvent = 
{
    OnConnect : {Code : 1},
    OnStore : {Code : 2},
    OnClientConnect : {Code : 100},
    OnServerConnect : {Code : 101},
    OnDataSideConnect : {Code : 102},
    OnClientDisconnect : {Code : 200},
    OnServerDisconnect : {Code : 201},
    OnDataSideDisconnect : {Code : 202},
    OnClientMessage : {Code : 300}, 
    OnServerMessage : {Code : 301}, 
    OnDataSideMessage : {Code : 302}, 
    OnClientKeepAlive : {Code : 400}, 
    OnServerKeepAlive : {Code : 401},
    OnDataSideKeepAlive : {Code : 402},
    OnClientPreviousMessage : {Code : 500},
    OnServerPreviousMessage : {Code : 501},
    OnDataSidePreviousMessage : {Code : 502}
}

var WebSocketMessageType = 
{
    Text : 0,
    Binary : 1,
    Close : 2
}

function WebSocketMessage(eventCode, messageType, message)
{
    this.eventCode = eventCode;
    this.MessageType = messageType;
    this.Message = message; 
    this.Length = this.MessageType == 0 ? new TextEncoder('utf-8').encode(message).length : message.length;
}


function MyWebSocket(url)
{
    this.websocket = new WebSocket(url);
    this.eventHandlers = [];
    this.OnMessageHandler = OnMessageHandler;
    this.SendMessages = SendMessages;
    this.AddEventHandler = (wsmEvent, Handler) => 
    {
        this.eventHandlers[wsmEvent.Code] = Handler;
    }
    this.websocket.onopen = ev => 
    {
        console.log("The WebSocket Connection Opened");
        SendMessages(this.websocket, [new WebSocketMessage(WebSocketMessageEvent.OnClientConnect , 0, "Hello")]);
    }

    this.websocket.onclose = ev =>
         console.log("The WebSocket Connection Closed");
    this.websocket.onerror = ev =>
         console.error(`An Error Has Occured When ${ev.type}`);
    this.websocket.onmessage = ev =>
         OnMessageHandler(ev, this.websocket, this.eventHandlers);
    
    return this;
}


async function OnMessageHandler(ev, websocket, eventHandlers) 
{
    var bytes = await ev.data.arrayBuffer();
    var wsMessages = [];
    var byteOffSet = 0;
    var isEffective, eventCode, messageType, length, message, wsMessage;
    do
    {
        isEffective = new DataView(bytes, byteOffSet, 1).getUint8(0);
        if(isEffective != IS_EFFECTIVE_CODE || bytes.length < HEADER_LENGTH)
        {
            console.error("The Stream Of ByteArray Is Not A Effective Message");
        }
        eventCode = new DataView(bytes,byteOffSet + MESSAGE_EVENT_CODE_LOCATION, UINT32_LENGTH).getInt32(0, true);
        messageType = new DataView(bytes,byteOffSet + MESSAGE_TYPE_LOCATION, UINT8_LENGTH).getUint8(0);
        length = new DataView(bytes,byteOffSet + LENGTH_LOCATION, UINT32_LENGTH).getInt32(0, true);
        message = new DataView(bytes,byteOffSet + MESSAGE_LOCATION, length)
        if(messageType == WebSocketMessageType.Text)
        {
            message = new TextDecoder("utf-8").decode(message);
        }
        
        wsMessage = new WebSocketMessage({Code : eventCode}, messageType, message);
        wsMessages.push(wsMessage);
        byteOffSet += (length + HEADER_LENGTH);
    }while(byteOffSet < bytes.byteLength);

    var sendMessage = await eventHandlers[eventCode](wsMessages);

    await SendMessages(websocket, sendMessage);



}
/// <summary>
/// 由`WebSocketMessage` 对象转换为 字节数组.
/// {
/// 基于websocket协议的自定义协议 : 
///     
///     byte 0 <==> IsEffective ? The Effective Code Is `200`
///     byte 1 ~ 4 <==> WebSocketMessage.eventCode.Code (int32) little endian
///     byte 5 <==> WebSocketMessage.MessageType (enum = 3)
///     byte 6 ~ 9 <==> WebSocketMessage.Length (int32)
///     byte 10 ~ 10 + length <==> WebSocketMessage.Message
/// }
/// </summary>
/// <param name="bytes"></param>
/// <returns></returns>
function SetAMessage(bytes, startIndex, wsMessage) 
{

    new DataView(bytes, startIndex + IS_EFFECTIVE_LOCATION, UINT8_LENGTH).setUint8(0, 200);
    new DataView(bytes, startIndex + MESSAGE_EVENT_CODE_LOCATION, UINT32_LENGTH).setInt32(0, wsMessage.eventCode.Code, true);
    new DataView(bytes, startIndex + MESSAGE_TYPE_LOCATION, UINT8_LENGTH).setUint8(0, wsMessage.MessageType);
    new DataView(bytes, startIndex + LENGTH_LOCATION, UINT32_LENGTH).setInt32(0, wsMessage.Length, true);
    var messageSegment = new DataView(bytes, startIndex + MESSAGE_LOCATION, wsMessage.Length);
    if (wsMessage.MessageType == WebSocketMessageType.Text)//Text
    {
        var messageStr = new TextEncoder("utf-8").encode(wsMessage.Message);
        for (var i = 0; i < wsMessage.Length; i++) {
            messageSegment.setUint8(i, messageStr[i]);
        }
    }
    else if (wsMessage.MessageType == WebSocketMessageType.Binary)
        for (var i = 0; i < wsMessage.Length; i++) {
            messageSegment.setUint8(i, wsMessage.Message[i]);
        }

    return wsMessage.Length + HEADER_LENGTH;


}

function SendMessages(websocket, wsMessages)
{
    return new Promise(resolve => 
    {
        var totalLength = 0;
        wsMessages.forEach(wsMessage => {
            totalLength += wsMessage.Length + HEADER_LENGTH;
        });
        var bytes = new ArrayBuffer(totalLength);
        var byteOffset = 0;
        wsMessages.forEach(wsMessage => {
            byteOffset += SetAMessage(bytes, byteOffset, wsMessage);
        });
        if (websocket.readyState == WebSocket.OPEN)
            websocket.send(bytes);
        resolve(0);
    });
    
}

export default
{
    MyWebSocket, WebSocketMessage, WebSocketMessageType, WebSocketMessageEvent
}

File: AndroidKotlinApi/HttpTest/app/src/main/java/com/example/httptest/mykits/MyWebSocket.kt
package com.example.httptest.mykits


import android.net.http.HttpResponseCache.install
import android.util.Log
import androidx.collection.arrayMapOf
import okhttp3.OkHttpClient
import okhttp3.Response
import okhttp3.WebSocket
import okio.ByteString
import okio.ByteString.Companion.toByteString
import java.lang.Exception
import java.nio.Buffer
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.Charset
import java.util.*
import kotlin.collections.ArrayList
import kotlin.experimental.or
import kotlin.properties.Delegates


class MyWebSocket(url: String) {
    companion object {

        const val IS_EFFECTIVE_CODE: UByte = 200u
        const val HEADER_LENGTH = 10

        const val UINT32_LENGTH = 4
        const val UINT8_LENGTH = 1

        const val IS_EFFECTIVE_LOCATION = 0
        const val MESSAGE_EVENT_CODE_LOCATION = 1

        const val MESSAGE_TYPE_LOCATION = 5
        const val LENGTH_LOCATION = 6
        const val MESSAGE_LOCATION = 10

    }

    data class WebSocketMessageEvent(val Code: Int) {
        companion object {
            val OnConnect = WebSocketMessageEvent(1)
            val OnStore = WebSocketMessageEvent(2)
            val OnClientConnect = WebSocketMessageEvent(100)
            val OnServerConnect = WebSocketMessageEvent(101)
            val OnDataSideConnect = WebSocketMessageEvent(102)
            val OnClientDisconnect = WebSocketMessageEvent(200)
            val OnServerDisconnect = WebSocketMessageEvent(201)
            val OnDataSideDisconnect = WebSocketMessageEvent(202)
            val OnClientMessage = WebSocketMessageEvent(300)
            val OnServerMessage = WebSocketMessageEvent(301)
            val OnDataSideMessage = WebSocketMessageEvent(302)
            val OnClientKeepAlive = WebSocketMessageEvent(400)
            val OnServerKeepAlive = WebSocketMessageEvent(401)
            val OnDataSideKeepAlive = WebSocketMessageEvent(402)
            val OnClientPreviousMessage = WebSocketMessageEvent(500)
            val OnServerPreviousMessage = WebSocketMessageEvent(501)
            val OnDataSidePreviousMessage = WebSocketMessageEvent(502)
        }
    }

    enum class WebSocketMessageType {
        Text,
        Binary,
        Close
    }


    data class WebSocketMessage(
        val EventCode: WebSocketMessageEvent,
        val MessageType: WebSocketMessageType,
        val Message: Any
    ) {

        var Length: Int = 0

        init {
            try {
                Length = when (this.MessageType) {
                    WebSocketMessageType.Text -> (Message as String).toByteArray().size
                    WebSocketMessageType.Binary -> (Message as ByteArray).size
                    else -> (Message as ByteArray).size
                }
            } catch (e: Exception) {
                Log.e(
                    toString(),
                    "Failed To Init The WebSocketMessage Class \n Caused By : ${e.message}"
                )
                e.printStackTrace()
            }

        }

    }

    val Url = url
    private val websocket = OkHttpWebSocket(Url)
    private val eventHandlers =
        arrayMapOf<Int, (ArrayList<WebSocketMessage>) -> ArrayList<WebSocketMessage>>()

    fun AddEventHandler(
        wsmEvent: WebSocketMessageEvent,
        Handler: (ArrayList<WebSocketMessage>) -> ArrayList<WebSocketMessage>
    ) {
        this.eventHandlers[wsmEvent.Code] = Handler
    }

    init {
        websocket.OnOpen = {
            try {
                Log.i(toString(), "The WebSocket Connection Opened");
                SendMessages(
                    arrayListOf<WebSocketMessage>(
                        WebSocketMessage(
                            WebSocketMessageEvent.OnClientConnect,
                            WebSocketMessageType.Text,
                            "Hello"
                        )
                    )
                )
            } catch (e: Exception) {
                Log.e(
                    toString(),
                    "Failed To Init MyWebSocket And Establish Connection \n Caused By : $e : ${e.message}"
                )
                e.printStackTrace()
            }

        }

        websocket.OnClosed = { code: Int, reason: String ->
            Log.i(toString(), "The WebSocket Connection Closed")
        }

        websocket.OnFaliure = { t: Throwable, response: Response? ->
            Log.e(toString(), "An Error Has Occured Caused By : $t : ${t.message}")
        }

        websocket.OnMessage = { bytes: ByteString ->
            try {
                OnMessageHandler(bytes.toByteArray())
            }catch (e : Exception) {
                Log.e(toString(),"An Exception Occured While Executing The OnMessageHandler \n Caused By : $e : ${e.message}")
                e.printStackTrace()
            }

        }

    }


    private fun OnMessageHandler(bytes: ByteArray) {
        val wsMessages = arrayListOf<WebSocketMessage>()
        var byteOffSet = 0
        var isEffective: UByte
        var eventCode: Int
        var messageType: WebSocketMessageType
        var length: Int
        var message: Any = byteArrayOf()
        var wsMessage: WebSocketMessage
        do {
            val wrapedBytes = ByteBuffer.wrap(bytes)
            isEffective = bytes[byteOffSet].toUByte()
            if (isEffective != IS_EFFECTIVE_CODE || bytes.lastIndex < HEADER_LENGTH) {
                Log.e(toString(), "The Stream Of ByteArray Is Not A Effective Message");
            }
            eventCode =
                ByteBuffer.wrap(bytes, byteOffSet + MESSAGE_EVENT_CODE_LOCATION, UINT32_LENGTH)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .int
            val wsMessageTypeCode = bytes[byteOffSet + MESSAGE_TYPE_LOCATION].toUByte()

            messageType = when (wsMessageTypeCode) {
                UByte.MIN_VALUE.or(0u) -> WebSocketMessageType.Text
                UByte.MIN_VALUE.or(1u) -> WebSocketMessageType.Binary
                else -> WebSocketMessageType.Close
            }.also {
                length =
                    ByteBuffer.wrap(bytes, byteOffSet + LENGTH_LOCATION, UINT32_LENGTH)
                        .order(ByteOrder.LITTLE_ENDIAN)
                        .int
                when (it) {
                    WebSocketMessageType.Text -> {
                        message = String(
                            ByteBuffer.wrap(bytes, byteOffSet + MESSAGE_LOCATION, length).toByteString().toByteArray(),

                        )
                    }
                    WebSocketMessageType.Binary -> {
                        message =
                            ByteBuffer.wrap(bytes, byteOffSet + MESSAGE_LOCATION, length).toByteString().toByteArray()
                    }

                }
            }


            wsMessage = WebSocketMessage(WebSocketMessageEvent(eventCode), messageType, message)
            wsMessages.add(wsMessage)
            byteOffSet += (length + HEADER_LENGTH)
        } while (byteOffSet <= bytes.lastIndex)

        val sendMessage = eventHandlers[eventCode]?.let { it(wsMessages) }

        if (sendMessage != null) {
            SendMessages(sendMessage)
        }


    }

    ///// <summary>
///// 由`WebSocketMessage` 对象转换为 字节数组.
///// {
///// 基于websocket协议的自定义协议 :
/////
/////     byte 0 <==> IsEffective ? The Effective Code Is `200`
/////     byte 1 ~ 4 <==> WebSocketMessage.eventCode.Code (int32) little endian
/////     byte 5 <==> WebSocketMessage.MessageType (enum = 3)
/////     byte 6 ~ 9 <==> WebSocketMessage.Length (int32)
/////     byte 10 ~ 10 + length <==> WebSocketMessage.Message
///// }
///// </summary>
///// <param name="bytes"></param>
///// <returns></returns>
    private fun SetAMessage(bytes: ByteArray, startIndex: Int, wsMessage: WebSocketMessage): Int {
        val wsMessageTypeCode: UByte = when (wsMessage.MessageType) {
            WebSocketMessageType.Text -> 0u
            WebSocketMessageType.Binary -> 1u
            else -> 2u
        }

        val isEffeciveSegment = ByteBuffer.wrap(
            bytes,
            startIndex + IS_EFFECTIVE_LOCATION,
            UINT8_LENGTH
        )
        val eventCodeSegment = ByteBuffer.wrap(
            bytes,
            startIndex + MESSAGE_EVENT_CODE_LOCATION,
            UINT32_LENGTH
        )
        val messageTypeSegment = ByteBuffer.wrap(
            bytes,
            startIndex + MESSAGE_TYPE_LOCATION,
            UINT8_LENGTH
        )
        val lengthSegment = ByteBuffer.wrap(
            bytes,
            startIndex + LENGTH_LOCATION,
            UINT32_LENGTH
        )
        val messageSegment = ByteBuffer.wrap(bytes, startIndex + MESSAGE_LOCATION, wsMessage.Length)
        isEffeciveSegment.put(IS_EFFECTIVE_CODE.toByte())
        eventCodeSegment.order(ByteOrder.LITTLE_ENDIAN).putInt(wsMessage.EventCode.Code)
        messageTypeSegment.put(wsMessageTypeCode.toByte())
        lengthSegment.order(ByteOrder.LITTLE_ENDIAN).putInt(wsMessage.Length)
        if (wsMessage.MessageType == WebSocketMessageType.Text)//Text
        {
            val messageStr = wsMessage.Message as String

            messageSegment.put(messageStr.toByteArray())

        } else if (wsMessage.MessageType == WebSocketMessageType.Binary) {
            messageSegment.put(wsMessage.Message as ByteArray)
        }

        return wsMessage.Length + HEADER_LENGTH;


    }

    fun SendMessages(wsMessages: ArrayList<WebSocketMessage>) {

        var totalLength = 0;
        wsMessages.forEach {
            totalLength += it.Length + HEADER_LENGTH
        }
        val bytes = ByteArray(totalLength)
        var byteOffset = 0
        wsMessages.forEach {
            byteOffset += SetAMessage(bytes, byteOffset, it);
        }

        websocket.Send(ByteBuffer.wrap(bytes).toByteString())

    }
}